crates/trie/src/state.rs

```diff
@@ -42,26 +50,74 @@ impl HashedPostState {
     pub fn from_bundle_state<'a>(
         state: impl IntoIterator<Item = (&'a Address, &'a BundleAccount)>,
     ) -> Self {
-        let mut hashed_state = Self::default();
+        let mut this = Self::default();
 
         for (address, account) in state {
             let hashed_address = keccak256(address);
-            if let Some(account) = &account.info {
-                hashed_state.insert_account(hashed_address, into_reth_acc(account.clone()))
-            } else {
-                hashed_state.insert_destroyed_account(hashed_address);
-            }
+            this.insert_account(hashed_address, account.info.clone().map(into_reth_acc));
 
             // insert storage.
             let mut hashed_storage = HashedStorage::new(account.status.was_destroyed());
 
             for (key, value) in account.storage.iter() {
                 let hashed_key = keccak256(B256::new(key.to_be_bytes()));
-                hashed_storage.insert_storage(hashed_key, value.present_value);
+                hashed_storage.insert_slot(hashed_key, value.present_value);
+            }
+            this.insert_hashed_storage(hashed_address, hashed_storage)
+        }
+        this.sorted()
+    }
+
+    /// Initialize [HashedPostState] from revert range.
+    /// Iterate over state reverts in the specified block range and
+    /// apply them to hashed state in reverse.
+    ///
+    /// NOTE: In order to have the resulting [HashedPostState] be a correct
+    /// overlay of the plain state, the end of the range must be the current tip.
+    pub fn from_revert_range<TX: DbTx>(
+        tx: TX,
+        range: RangeInclusive<BlockNumber>,
+    ) -> Result<Self, DatabaseError> {
+        let mut state =
+            HashMap::<Address, (Option<Option<Account>>, HashMap<B256, U256>)>::default();
+
+        // Iterate over account changesets in reverse.
+        let account_changesets = tx
+            .cursor_read::<tables::AccountChangeSet>()?
+            .walk_range(range.clone())?
+            .collect::<Result<Vec<_>, _>>()?;
+        for (_, account_before) in account_changesets.into_iter().rev() {
```

Comment:
Rather than collecting into a `Vec` and then iterating in reverse you can just iterate forward and only take the first value you see, e.g.

```rust
for changeset in tx.
    .cursor_read::<tables::AccountChangeSet>()?
    .walk_range(range.clone())? {
    let (_, account_before) = changeset?;
    let AccountBeforeTx { address, info } = account_before;
    if let Entry::Vacant(entry) = state.entry(address) {
        entry.insert(Some(info))
    }
}
```

Reply:
good catch, ty